README
megatools - command line client application for Mega
====================================================

Megatools is a collection of programs for accessing Mega service from 
a command line of your desktop or server.

Megatools allow you to copy individual files as well as entire directory 
trees to and from the cloud. You can also perform streaming downloads for 
example to preview videos and audio files, without needing to download 
the entire file.

Megatools are robust and optimized for fast operation - as fast as Mega 
servers allow. Memory requirements and CPU utilization are kept at minimum.

You can register account using a 'megareg' tool, with the benefit of having 
true control of your encryption keys.

Mega website can be found at http://mega.nz.

Megatools can be downloaded at http://megatools.megous.com


Tools
=====

  megareg      Register and verify a new mega account
  megadf       Show your cloud storage space usage/quota
  megals       List all remote files
  megamkdir    Create remote directory
  megarm       Remove remote file or directory
  megaput      Upload individual files
  megaget      Download individual files
  megadl       Download file from a "public" Mega link
               (doesn't require login)
  megastream   Streaming download of a file
               (can be used to preview videos or music)
  megacopy     Upload or download a directory tree
  megafs       Mount remote filesystem locally.


All of these tools do:

- Local caching of remote session/filesystem information 
  for faster execution. Cache is encrypted with your password 
  key.
- Support loading login credentials from a configuration file


Usage
=====

See man pages for how to use individual tools and how to configure them:

  man megatools


Windows Quirks
==============

On Windows, if you see weird characters in your megals output, you'll need
to set correct CHARSET environment variable. For example on Czech Windows 
this would mean executing this command in cmd before using the tools:

  set CHARSET=CP852

This is just a cosmetic issue. Internally, megatools always work with UTF-8
file names, and even if the tool's terminal output is corrupted, files names
of downloaded/uploaded files will be correct.


Installation on Mac OS X (Mountain Lion)
========================================

Thanks to Carl Moden, megatools is available in Homebrew (http://brew.sh/).

You can therefrore install megatools with:

  brew install megatools

To install megatools by hand you need to:

1. Install glib-networking with a ca-bundle:

  export BUNDLE=/usr/local/share/ca-bundle.crt
  curl http://curl.haxx.se/ca/cacert.pem -o $BUNDLE
  /usr/bin/sed -ie "s|--without-ca-certificates|--with-ca-certificates=$BUNDLE|" /usr/local/Library/Formula/glib-networking.rb
  brew install glib-networking

2. Make the pkgconfig (.pc) file for libcurl findable by megatools configure script:

  export PKG_CONFIG_PATH=/usr/local/Library/ENV/pkgconfig/10.8

3. Untar megatools, run `./configure`, `make` and `make install` as usual.


Installation on Gentoo
======================

Third-party ebuild is available at:

  https://github.com/megabaks/stuff/tree/master/app-misc/megatools


Installation on Debian
======================

Megatools is available in sid thanks to Alberto Garcia:

  https://packages.debian.org/unstable/megatools


Installation on FreeBSD
=======================

Megatools is available in freshports thanks to Maxim V. Kostikov:

  http://www.freshports.org/net/megatools/


Building megatools from source code
===================================

You will need to install dependnencies first, to be able to do the build from
the official source code tarball, which is available only at 
http://megatools.megous.com.

You can also check that the code was released by me, using GPG signatures
provided alongside the code.

Please don't report build issues against code that you downloaded from github. 
These reports will be closed without explanation, with reference to this README
file. It is not expected to work on all distributions. Use the official tarball.


On Debian, Ubuntu:

  apt-get -y install build-essential libglib2.0-dev libssl-dev libcurl4-openssl-dev libgirepository1.0-dev

On Fedora:

  yum -y install gcc make glib2-devel libcurl-devel openssl-devel gmp-devel tar

On OpenSUSE:

  zypper -n install gcc make glib2-devel libcurl-devel openssl-devel gmp-devel

On Arch Linux:

  pacman -Sy --noconfirm --needed pkg-config gcc make glib2 curl gmp nettle


Author
======

Megatools were written by Ondřej Jirman <megous@megous.com>, 2014
Webiste: http://megatools.megous.com


Support and bug reports
=======================

If you think you've found bug in megatools, send a report including enough
information for recreating the issue to: megous@megous.com


License
=======

Megatools are licensed under GPLv2 with OpenSSL exemption, see LICENSE
file for details.

This product includes software developed by the OpenSSL Project for use
in the OpenSSL Toolkit. (http://www.openssl.org/)


NEWS
megatools 1.9.97 - 2016-02-02
=============================

Bugfix release.

Fixes:
- Mega started compressing HTTP responses to API calls. Megatools now
  uses libcurl to handle API requests, so HTTP compression is now 
  supproted.

megatools 1.9.96 - 2016-01-02
=============================

Bugfix release.

Fixes:
- Refer to mega.nz and use mega.nz links instead of deprecated mega.co.nz
- Fix various build issues

megatools 1.9.95 - 2015-04-15
=============================

Bugfix and optimization release.

Fixes:
- Improve compatibility wit older curl (whatever version is in centos 6.5)
- Fix regression in megadl: fails to download exported folders #90
- Remove megamv as it is not and will not be implemented in 1.9 branch
- Slow down progress reporting on windows to once per second #102
- Correct the default values in mega_session_load()

Improvements:
- Rename megasync to megacopy to avoid conflict with official mega sync client
- Optimize session loading - issue #100
- Optimize session loading - further optimization - issue #100
- There's no need to use dbus-launcher when running megatools outside running
  desktop to avoid a warning message.

megatools 1.9.94 - 2015-01-02
=============================

Bugfix and optimization release.

Fixes:
- Force use of IPv4 addresses
- Fix path sanitization so that trailing slashes are ignored
- Document config file oddities in megarc manpage
- Add info on how to build/install megatools on various distros

Improvements:
- Optimize path handling code so that filesystems with many files are 
  usable. Previously, megatools would hang for quite some time when
  updating it's path map.
- Use non-recursive makefile to build megatools

megatools 1.9.93 - 2014-10-24
=============================

Added OpenSSL exemption to the license.

Fixes:

- megadf now returns correct values
- fixed path sanitization in mega_session_stat
- raised safety limit os size of POST responses from 32 to 256 MiB
- show progress when downloading folder using megadl
- optimized buffer allocation during donwload and upload
- report details of errors when stream callback fails to read/write file
- fix detection of convert executeable for preview generation
- allow to disable file previews

megatools 1.9.92 - 2014-08-24
=============================

Created stable branch for bugfixes. This is the release from the stable 
branch.

This release contains mostly small bug fixes. Most important change is
that Megatools now recognizes server busy errors as reported by mega.nz
via HTTP status 500 Server Is Busy, and retries the communication. This
reduces command failures by a lot, and you'll not see 500 Server Is Busy 
errors anymore.

Another major news is update to the windows package. There are now two 
packages for 32bit and 64bit windows platform, and all dlls were updated to
their latest versions.

Other news:

- Added 64bit windows package
- Unmask Megatools to Mega.nz (user Megatools/VERSION User-Agent)
- Check for a2x if --enable-docs-build=yes
- Megadl now reports file download errors
- Add megadl --print-names option
- Fix megasync error (issue #45 on github)
- Fix megadl used with pipe (--path -) on windows (issue #38)
- Installation instructions for Mac
- Add info about Gentoo ebuild to the README file (issue #26)
- Fix Megatools thumbnails for PDF (issue #5)
- Do repeated API calls after 500 Server Is Busy and increase retry delay
- Fix issue to download a file with accent with megaget (issue #22)

megatools 1.9.91 - 2013-04-14
=============================

Please note, that megatools is using GNOME release versioning. That means
that in MAJOR.MINOR.PATCH, odd MINOR number signifies development version
of the tools. PATCH version is increased on each public release.

By using odd MINOR versions of the software you are using an unstable
branch. That doesn't mean it will be crashing on you, but it may have 
certain quirks, and oddities, missing or incomplete features, etc.

Next stable release is going to be 2.0.0.

To the news:

- Added man pages for all tools
- Removed megastream in favor of mega{dl,get} --path -
- Megadf - fix unsigned overflow when server reports 0 total space
- Implemented --debug option
- Win32: Drop requirement for manual setup of GIO_EXTRA_MODULES
- Load certificates from the binaries folder on windows
- Enable CA validation for TLS connections
- Fix compilation on FreeBSD by adding --enable-warnings
- Add note that on FreeBSD GNU make must be used to build the tools
- Fix typedef redefinitions (error on older gcc's)
- Fix megals --names logic
- megaput/megaget/megadl/megasync: --no-progress support
- Fix mega_session_dl streaming mode segfault
- megamkdir/megarm: Check for required arguments
- Pretty print JSON in debug output

AUR package is available at https://aur.archlinux.org/packages/megatools/

Ubunut PPA is availble at https://launchpad.net/~megous/+archive/ppa

If someone wants to take over packaging for Ubuntu or Debian, that would
be very appretiated, as my speciality is Arch Linux.


megatools 1.9.90 - 2013-04-11
=============================

Implemented much simpler and faster json parser than json-glib, which was
used before. Megatools now use sjson, which you can find at:

  https://github.com/megous/sjson

Json-glib dependency was thus dropped.

Initial work on getting rid of dependency on libcurl.

Initial work on the new libmega library with support for gobject-introspection
based bindings to various languages: JavaScript, Python, Ruby, LUA,...

Experimental implementation of Mega API in JavaScript (gjs), this should help
people develop GNOME GUI for Mega in the future.

New command line tools:

  megafs  - mount mega.nz filesystem usong FUSE (Linux only)
  megamv  - move and rename remote files and directories


megatools 1.2.2 - 2013-04-11
============================

Bugs fixed:

- Trash and Inbox paths are no longer mixed up (bug reported by Ovidiu)
- megadl segfault

tools/dl.c
/*
 *  megatools - Mega.nz client library and tools
 *  Copyright (C) 2013  Ondřej Jirman <megous@megous.com>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License along
 *  with this program; if not, write to the Free Software Foundation, Inc.,
 *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include "tools.h"
#ifdef G_OS_WIN32
#include <io.h>
#include <fcntl.h>
#endif

static gchar* opt_path = ".";
static gboolean opt_stream = FALSE;
static gboolean opt_noprogress = FALSE;
static gboolean opt_print_names = FALSE;

static GOptionEntry entries[] =
{
  { "path",          '\0',   0, G_OPTION_ARG_FILENAME,  &opt_path,  "Local directory or file name, to save data to",  "PATH" },
  { "no-progress",   '\0',   0, G_OPTION_ARG_NONE,    &opt_noprogress,  "Disable progress bar",   NULL},
  { "print-names",   '\0',   0, G_OPTION_ARG_NONE,    &opt_print_names,  "Print names of downloaded files",   NULL},
  { NULL }
};

static gchar* cur_file = NULL;
static mega_session* s;

static gboolean status_callback(mega_status_data* data, gpointer userdata)
{
  if (opt_stream && data->type == MEGA_STATUS_DATA)
  {
    fwrite(data->data.buf, data->data.size, 1, stdout);
    fflush(stdout);
  }

  if (data->type == MEGA_STATUS_FILEINFO)
  {
    cur_file = g_strdup(data->fileinfo.name);
  }

  if (!opt_noprogress && data->type == MEGA_STATUS_PROGRESS)
  {
    gc_free gchar* done_str = g_format_size_full(data->progress.done, G_FORMAT_SIZE_IEC_UNITS);
    gc_free gchar* total_str = g_format_size_full(data->progress.total, G_FORMAT_SIZE_IEC_UNITS);

    if (data->progress.total > 0)
      g_print(ESC_WHITE "%s" ESC_NORMAL ": " ESC_GREEN "%" G_GUINT64_FORMAT "%%" ESC_NORMAL " - " ESC_GREEN "%s" ESC_NORMAL " of %s" ESC_CLREOL "\r", cur_file, 100 * data->progress.done / data->progress.total, done_str, total_str);
  }

  return FALSE;
}

// download operation

static gboolean dl_sync_file(mega_node* node, GFile* file, const gchar* remote_path)
{
  gc_error_free GError *local_err = NULL;
  gc_free gchar* local_path = g_file_get_path(file);

  if (g_file_query_exists(file, NULL))
  {
    g_printerr("ERROR: File already exists at %s\n", local_path);
    return FALSE;
  }

  if (!opt_noprogress)
    g_print("F %s\n", local_path);

  if (!mega_session_get(s, local_path, remote_path, &local_err))
  {
    if (!opt_noprogress)
      g_print("\r" ESC_CLREOL);
    g_printerr("ERROR: Download failed for %s: %s\n", remote_path, local_err->message);
    return FALSE;
  }

  if (!opt_noprogress)
    g_print("\r" ESC_CLREOL);

  if (opt_print_names)
    g_print("%s\n", local_path);

  return TRUE;
}

static gboolean dl_sync_dir(mega_node* node, GFile* file, const gchar* remote_path)
{
  gc_error_free GError *local_err = NULL;
  gc_free gchar* local_path = g_file_get_path(file);

  if (!g_file_query_exists(file, NULL))
  {
    if (!opt_noprogress)
      g_print("D %s\n", local_path);

    if (!g_file_make_directory(file, NULL, &local_err))
    {
      g_printerr("ERROR: Can't create local directory %s: %s\n", local_path, local_err->message);
      return FALSE;
    }
  }
  else
  {
    if (g_file_query_file_type(file, G_FILE_QUERY_INFO_NOFOLLOW_SYMLINKS, NULL) != G_FILE_TYPE_DIRECTORY)
    {
      g_printerr("ERROR: Can't create local directory %s: file exists\n", local_path);
      return FALSE;
    }
  }

  // sync children
  GSList* children = mega_session_get_node_chilren(s, node), *i;
  for (i = children; i; i = i->next)
  {
    mega_node* child = i->data;
    gc_free gchar* child_remote_path = g_strconcat(remote_path, "/", child->name, NULL);
    gc_object_unref GFile* child_file = g_file_get_child(file, child->name);

    if (child->type == 0)
    {
      dl_sync_file(child, child_file, child_remote_path);
    }
    else
    {
      dl_sync_dir(child, child_file, child_remote_path);
    }
  }

  g_slist_free(children);
  return TRUE;
}

int main(int ac, char* av[])
{
  gc_error_free GError *local_err = NULL;
  gc_regex_unref GRegex *file_regex = NULL, *folder_regex = NULL;
  gint i;
  int status = 0;

  tool_init_bare(&ac, &av, "- download exported files from mega.nz", entries);

  if (!strcmp(opt_path, "-"))
  {
    opt_noprogress = opt_stream = TRUE;

    // see https://github.com/megous/megatools/issues/38
#ifdef G_OS_WIN32
    setmode(fileno(stdout), O_BINARY);
#endif
  }

  if (ac < 2)
  {
    g_printerr("ERROR: No links specified for download!\n");
    tool_fini(NULL);
    return 1;
  }

  if (opt_stream && ac != 2)
  {
    g_printerr("ERROR: Can't stream from multiple files!\n");
    tool_fini(NULL);
    return 1;
  }

  // prepare link parsers

  file_regex = g_regex_new("^https?://mega(?:\\.co)?\\.nz/#!([a-z0-9_-]{8})!([a-z0-9_-]{43})$", G_REGEX_CASELESS, 0, NULL);
  g_assert(file_regex != NULL);

  folder_regex = g_regex_new("^https?://mega(?:\\.co)?\\.nz/#F!([a-z0-9_-]{8})!([a-z0-9_-]{22})$", G_REGEX_CASELESS, 0, NULL);
  g_assert(folder_regex != NULL);

  // create session

  s = mega_session_new();

  mega_session_watch_status(s, status_callback, NULL);

  // process links
  for (i = 1; i < ac; i++)
  {
    gc_match_info_unref GMatchInfo* m1 = NULL;
    gc_match_info_unref GMatchInfo* m2 = NULL;
    gc_free gchar* key = NULL;
    gc_free gchar* handle = NULL;
    gc_free gchar* link = tool_convert_filename(av[i], FALSE);

    if (g_regex_match(file_regex, link, 0, &m1))
    {
      handle = g_match_info_fetch(m1, 1);
      key = g_match_info_fetch(m1, 2);

      // perform download
      if (!mega_session_dl(s, handle, key, opt_stream ? NULL : opt_path, &local_err))
      {
        if (!opt_noprogress)
          g_print("\r" ESC_CLREOL "\n");
        g_printerr("ERROR: Download failed for '%s': %s\n", link, local_err->message);
        g_clear_error(&local_err);
        status++;
      }
      else
      {
        if (!opt_noprogress)
          g_print("\r" ESC_CLREOL "Downloaded %s\n", cur_file);
        if (opt_print_names)
          g_print("%s\n", cur_file);
      }
    }
    else if (g_regex_match(folder_regex, link, 0, &m2))
    {
      if (opt_stream)
      {
        g_printerr("ERROR: Can't stream from a directory!\n");
        tool_fini(s);
        return 1;
      }

      handle = g_match_info_fetch(m2, 1);
      key = g_match_info_fetch(m2, 2);

      // perform download
      if (!mega_session_open_exp_folder(s, handle, key, &local_err))
      {
        g_printerr("ERROR: Can't open folder '%s': %s\n", link, local_err->message);
        g_clear_error(&local_err);
      }
      else
      {
        mega_session_watch_status(s, status_callback, NULL);

        GSList* l = mega_session_ls(s, "/", FALSE);
        if (g_slist_length(l) == 1)
        {
          mega_node* root_node = l->data;

          gc_object_unref GFile* local_dir = g_file_new_for_path(opt_path);
          if (g_file_query_file_type(local_dir, G_FILE_QUERY_INFO_NOFOLLOW_SYMLINKS, NULL) == G_FILE_TYPE_DIRECTORY)
          {
            gc_free gchar* node_path = mega_node_get_path_dup(root_node);
            dl_sync_dir(root_node, local_dir, node_path);
          }
          else
          {
            g_printerr("ERROR: %s must be a directory\n", opt_path);
          }
        }
        else
        {
          g_printerr("ERROR: EXP folder fs has multiple toplevel nodes? Weird!\n");
        }

        g_slist_free(l);
      }
    }
    else
    {
      g_printerr("WARNING: Skipping invalid Mega download link: %s\n", link);
    }
  }

  tool_fini(s);
  return status;
}
